tests/MockStorage.js

```diff
@@ -180,8 +180,18 @@ export class MockStorage {
         }
       }
     }
-
+    const oldDoc = edv.documents.get(doc.id);
     edv.documents.set(doc.id, doc);
+    if(oldDoc) {
+      // if an indexed value was not updated
+      // it's attributes is an empty array.
+      // so we need to use the last attributes.
+      for(const index in doc.indexed) {
+        if(doc.indexed[index].attributes.length === 0) {
+          doc.indexed[index].attributes = oldDoc.indexed[index].attributes;
```

Comment:
I don't understand what's going on here. If the new doc doesn't have attributes we shouldn't be adding them in from an old document ... the new doc should be honored. This is an update where the attributes were removed from the new doc, right?

I would have expected this patch to go through the old doc's indexed attributes and remove it from all indexes... and then just run the rest of the code as normal, which would add the new doc to any indexed attributes (which may be none). *Perhaps* this change has that effect, but if so, it seems convoluted how it gets there.

Reply:
1. it's not a new doc it's an update to an existing doc.
2. the update does not change a property that is indexed.
3. because an indexed property is not changed no attributes are generated.

These were the cases I was considering in this patch:
Case 1:
1. User updates an existing document.
2. User's update changes an indexed property.
3. Update produces a new attributes array for the indexed field.
4. We check and do not overwrite the new attributes.
5. new attribute is indexed with latest document.

Case 2:
1. User updates an exiting document.
2. User's update does not change an indexed property.
3. The update passed to storage here will have an empty attributes array.
4. Because attributes is empty the indexed document is not updated (this is Orie's bug)
5. So we preserve the original attributes in our mock database resulting in an updated document.

I'm not aware of the ability to delete previous indexes however laying out the cases I do believe has revealed a minor new bug: if you change the attribute.value the key used to index changes which might result in the old unused key staying. While that does not result in test failure it does mean the test database preserves old keys. Although when I think about it not sure if mongo deletes the old indexed keys itself.

Basically this code is taking on the job of the database. I believe mongo is not destructive on update.

Reply:
> it's not a new doc it's an update to an existing doc.
the update does not change a property that is indexed.
because an indexed property is not changed no attributes are generated.

An update replaces an existing document entirely. There is no patch functionality. A document that is sent as an update that does not include a previously indexed field should cause that field to be dropped.

So, in short, I don't think those cases exist at all. There's no concept of a patch/partial update in the API, there is only total replacement.

Reply:
@dlongley I'm really sorry I totally forgot the behavior of this library even though I did use it before! I believe that orie's second test (`should have matching sequence numbers in indexes`) should be failing then because:

```
    client.ensureIndex({attribute: 'content.indexedKey'});
    const testId = await EdvClient.generateId();
    const doc = {
      id: testId,
      content: {someKey: 'someValue', indexedKey: 'value1'}
    };
    const version1 = await client.insert({
      doc,
      invocationSigner,
      keyResolver
    });
    version1.content = {someKey: 'aNewValue'};
    await client.update({doc: version1, invocationSigner, keyResolver});
```

then the second update to the same doc removes the `indexedKey`. 

`should only contain one indexed document after 3 updates`
keeps updates with the indexedKey in the doc.content so it should succeed.

`should have matching sequence numbers in indexes` however replaces the indexedKey with nothing so there should be no document in the old index hence it should be failing.

So basically if the indexed property is removed in an update, the indexes that store that document should also be removed. Or what you said in the first place which is just replace the index entirely.

Reply:
@aljones15, no worries! Yeah, that second update removes `indexedKey` so the doc should get dropped from the index. Reading the rest sounds like you've got a handle on things now. Thanks!

Reply:
@dlongley this now removes the indexed documents on update: https://github.com/digitalbazaar/edv-client/pull/18/commits/91e6ac4cf2d27970276c9e42eec616268dd7c13f